src/aggregate/aggregate_exec.c

```diff
@@ -356,23 +356,71 @@ static void blockedClientReqCtx_destroy(blockedClientReqCtx *BCRctx) {
 }
 
 void AREQ_Execute_Callback(blockedClientReqCtx *BCRctx) {
-  BCRctx->req->sctx->redisCtx = RedisModule_GetThreadSafeContext(BCRctx->blockedClient);
+  RedisModuleCtx *redis_ctx = BCRctx->req->sctx->redisCtx = RedisModule_GetThreadSafeContext(BCRctx->blockedClient);
   BCRctx->req->reqflags |= QEXEC_F_HAS_THCTX;
   // lockspec
   RedisSearchCtx_LockSpecRead(BCRctx->req->sctx);
-  AREQ_Execute(BCRctx->req, blockedClientReqCtx_getRedisctx(BCRctx));
+  QueryError status = {0};
+  if (prepareExecutionPlan(&BCRctx->req, &status) != REDISMODULE_OK) {
+    // Enrich the error message that was caught to include the fact that the query ran
+    // in a background thread.
+    QueryError detailed_status = {0};
+    QueryError_SetErrorFmt(&detailed_status, QueryError_GetCode(&status),
+                           "The following error was caught upon running the query asynchronously: %s", QueryError_GetError(&status));
+    QueryError_ClearError(&status);
+    QueryError_ReplyAndClear(redis_ctx, &detailed_status);
+    if (BCRctx->req) {
+      AREQ_Free(BCRctx->req);
+    }
+  } else {
+    AREQ_Execute(BCRctx->req, redis_ctx);
+  }
+
 
   // No need to unlock spec as AREQ_Execute calls ctx cleanup.
   RS_CHECK_FUNC(RedisModule_BlockedClientMeasureTimeEnd, BCRctx->blockedClient);
 
   blockedClientReqCtx_destroy(BCRctx);
 }
 
+int prepareExecutionPlan(AREQ **r, QueryError *status) {
+  int rc = REDISMODULE_ERR;
+  AREQ *req = *r;
+  RedisSearchCtx *sctx = req->sctx;
+  RSSearchOptions *opts = &req->searchopts;
+  QueryAST *ast = &req->ast;
+
+  ConcurrentSearchCtx_Init(sctx->redisCtx, &req->conc);
```

Comment:
I guess this was CP here? probably a dead code

Reply:
Well, I didn't change stuff, just moved...  `ConcurrentSearchCtx_Init(sctx->redisCtx, &req->conc);` was there originally. Is that redundant?

Reply:
lets keep it for now